docs/features/nullable-reference-types.md

```diff
@@ -43,8 +43,47 @@ A number of null checks affect the flow state when tested for:
   - `static bool object.Equals(object, object)`
   - `static bool object.ReferenceEquals(object, object)`
   - `bool object.Equals(object)` and overrides
-  - `bool IEquatable<T>(T)` and implementations
-  - `bool IEqualityComparer<T>(T, T)` and implementations
+  - `bool IEquatable<T>.Equals(T)` and implementations
+  - `bool IEqualityComparer<T>.Equals(T, T)` and implementations
+
+Some null checks are "pure null tests", which means that they can cause a variable whose flow state was previously not-null to update to maybe-null. Pure null tests include:
+- `x == null`, `x != null` *whether using a built-in or user-defined operator*
+- `(Type)x == null`, `(Type)x != null`
+- `x is null`
+- `object.Equals(x, null)`, `object.ReferenceEquals(x, null)`
+- `IEqualityComparer<Type?>.Equals(x, null)`
+
+All of the above checks except for `x is null` are commutative. For example, `null == x` is also a valid pure null test.
+
+Some expressions which may not return `bool` are also considered pure null tests:
+- `x?.Member` will change the receiver's flow state to maybe-null unconditionally
+- `x ?? y` will change the LHS expression's flow state to maybe-null unconditionally
+
+Example of how a pure null test can affect flow analysis:
+```cs
+string? s = "hello";
```

Comment:
>? [](start = 6, length = 1)

nit: I think these examples would be more clear removing the `?` from the definition.